Fix verify codecgen

lavalamp opened the pull request:
Fixes #25103.

lavalamp committed:
Fix codecgen

* Make verify call update
* Make update's topological sort not confused about the "io" package

lavalamp commented:
fyi @wojtek-t

mml commented:
@lavalamp :'(

```
FAILED   ./hack/../hack/verify-codecgen.sh	369s
Verifying ./hack/../hack/verify-codegen.sh
```

lavalamp commented:
Huh, I ran it multiple times locally. running again...

lavalamp commented:
Hm. My local system comes up with a different ordering than the remote system. That's probably the culprit.

lavalamp commented:
I have to go home. @wojtek-t please feel free to take this over if you see this before I get to it. I think the topological sort is not portable.

wojtek-t commented:
>
I have to go home. @wojtek-t please feel free to take this over if you see this before I get to it. I think the topological sort is not portable.

Why do you think so? It doesn't use anything super-fancy - it's just using some very standard commands.

Regarding the PR itself, this tricky. The files that are generated depends on whether the other ones (they potentially depend on) are generated or not. So we need a well-defined order of generation (the same for verification).

We should do in topological order, because then the gains are the biggest then.

lavalamp commented:
> Why do you think so?

Because I looked at the test logs and the logs my system uses, and files were processed in a different order. Perhaps because the files aren't deterministically sorted before the tsort, but I haven't verified that yet.

> We should do in topological order, because then the gains are the biggest then.

Nah, in the verify script it doesn't matter. You have to generate every file in the success case anyway, so you might as well just call the main script, like I've changed it to do.

mml commented:
> Perhaps because the files aren't deterministically sorted before the tsort

My money is here.  Each invocation may be correct, but it's not stable.

wojtek-t commented:
But it shouldn't matter - the topological sorting is supposed to fix that...

mml commented:
The tsort is supposed to make sure we do all dependencies of file F before doing F itself.  Consider only 3 files below:

```
B <----
      +--- A
C <----
```

There are two legitimate orders here: [C, B, A] and [B, C, A].  I haven't carefully read the tsort code, but AIUI which one gets generated depends on the order in which the files are emitted from find in the first place.

mml commented:
...and if that's correct, it should be enough to just do `generated_files=($(find ... | sort))`

wojtek-t commented:
@mml - yes that's correct, those two orders can be generated by the existing code.

However, what I was trying to say is that both of them should result in the same generated code according to what I know about ugorji. So it seems it's not the case.

>
...and if that's correct, it should be enough to just do generated_files=($(find ... | sort))

Agree. @lavalamp - can you try that (both in update and verify scripts)?

lavalamp commented:
Yeah, tsort is not deterministic if the things being sorted don't form a tree. That's what I thought was the problem, and I changed it. Actually the problem is the "io" package is confused with the testing type package under go2idl. I guess "io" is used or not depending on go version. Anyway, I fixed that.

lavalamp commented:
This is working now. @mml or @wojtek-t PTAL

mml commented on hack/verify-codecgen.sh:
```diff
@@ -35,85 +35,27 @@ generated_files=($(
-for (( i=0; i<number; i++ )); do
-  visited[${i}]=false
+for generated_file in ${generated_files[@]}; do
+  cat "${generated_file}" > "${generated_file}.original"
```
nit: why `cat` instead of `cp`?

lavalamp replied:
I didn't want to worry about cp barfing if the file already existed.

mml commented on hack/verify-codecgen.sh:
```diff
@@ -35,85 +35,27 @@ generated_files=($(
+  rm -f "${generated_file}.original"
 done
+
+if [[ $ret -ne 0 ]]; then
```
nit: Seems safe to drop this surrounding `if`.

lavalamp replied:
fixed

mml commented:
The e2e problem was fixed by #25159

luxas commented:
@k8s-bot e2e test this please github issue: #25159

wojtek-t commented:
LGTM - thanks!
